### PR TITLE
qcut: Option to return -inf/inf as lower/upper bound

### LINKED ISSUE #17282: qcut: Option to return -inf/inf as lower/upper bound
#### Code Sample

```python
>>> x = pd.qcut([2,3,4,5,6,7,8,9], q=5, duplicates='drop')
>>> x.categories
IntervalIndex([(1.999, 3.4], (3.4, 4.8], (4.8, 6.2], (6.2, 7.6], (7.6, 9.0]]
              closed='right',
              dtype='interval[float64]')

>>> pd.cut([1,5,6,10], x.categories)
[NaN, (4.8, 6.2], (4.8, 6.2], NaN]
Categories (5, interval[float64]): [(1.999, 3.4] < (3.4, 4.8] < (4.8, 6.2] < (6.2, 7.6] < (7.6, 9.0]]
```
#### Problem description

I'm currently using `qcut` and `cut` together for Machine Learning. I use `qcut` to cut training data into quantiles and use `cut` to cut the test data into the same bins.

However, if a value in the test data is too high/low, it will violate the Categories created by qcut, and the resulting category will be NaN. A solution to this is to create an option to return -inf/inf as lower/upper bound of the categories.

#### Expected Output
```python
>>> x = pd.qcut([2,3,4,5,6,7,8,9], q=5, duplicates='drop', inf_bounds=True)
>>> x.categories
IntervalIndex([(-inf, 3.4], (3.4, 4.8], (4.8, 6.2], (6.2, 7.6], (7.6, inf)]
              closed='right',
              dtype='interval[float64]')
>>> pd.cut([1,5,6,10], x.categories)
[(-inf, 3.4], (4.8, 6.2], (4.8, 6.2], (7.6, inf)]
Categories (5, interval[float64]): [(-inf, 3.4] < (3.4, 4.8] < (4.8, 6.2] < (6.2, 7.6] < (7.6, inf)]
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-514.26.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by dberenbaum**
- [x] closes #17282
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/reshape/tile.py [hunk 1]
```diff
@@ -262,6 +264,9 @@ def qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise'):
         The precision at which to store and display the bins labels
     duplicates : {default 'raise', 'drop'}, optional
         If bin edges are not unique, raise ValueError or drop non-uniques.
+    bounded : bool, optional
+        Use the min/max of the distribution as the lower/upper bounds if True,
+        otherwise use -inf/inf. Ignored if dtype is datetime/timedelta.
```
Comment (gfyoung) [inline pandas/core/reshape/tile.py]: Move this below the version added, and add your own for `bounded` i.e.:

~~~
   .. versionadded:: 0.20.0
bounded: bool, optional
...
   .. versionadded:: 0.24.0
~~~

File: pandas/core/reshape/tile.py [hunk 2]
```diff
@@ -302,6 +307,12 @@ def qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise'):
     else:
         quantiles = q
     bins = algos.quantile(x, quantiles)
+    if not bounded and not dtype:
+        if is_integer_dtype(bins):
+            bins = bins.astype(np.float64)
+        bins[0] = -np.inf
+        bins[-1] = np.inf
+        pass
```
Comment (gfyoung) [inline pandas/core/reshape/tile.py]: Why do we have a `pass` here?

File: pandas/tests/reshape/test_tile.py [hunk 3]
```diff
@@ -479,6 +479,14 @@ def test_cut_read_only(self, array_1_writeable, array_2_writeable):
         tm.assert_categorical_equal(cut(hundred_elements, array_1),
                                     cut(hundred_elements, array_2))
 
+    def test_qcut_unbounded(self):
+        labels = qcut(range(5), 4, bounded=False)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: Reference issue number as a comment above this line.

**Review by gfyoung — APPROVED** [review]
LGTM!

cc @jreback

**Comment by dberenbaum** [comment]
Checking in on this. How do I move this forward?

**Comment by gfyoung** [comment]
@dberenbaum : So sorry!  Didn't realize this one fell through the cracks.

cc @jreback @jorisvandenbossche @TomAugspurger : This one has been ready to go IMO for over a month.  Could one of you take a look?

**Comment by jreback** [comment]
I am not sure I like this, we are growing too many keywords here. Now we have 2 ways of specifying the bins. Is this really a problem to use np.inf in the bins specification?

**Comment by dberenbaum** [comment]
I'm not following how there are 2 ways of specifying the bins. I only see a way to specify quantiles in `qcut` (there is a bins arg in `cut`). I'm not sure how else to return `np.inf` as a bin edge using `qcut`.

The use cases I have are very similar to the one raised by @prcastro in #17282. When I need to 1) compare distributions or 2) bin numerical values into known categories, I use `qcut` to bin by quantiles and then bin the new distribution by those categories. This is problematic when the new distribution contains out of bounds values. The best alternative I've found is to avoid using `qcut`:

```
bins = np.quantile(x, [0, 0.25, 0.5, 0.75, 1])
bins[0] = -np.inf
bins[-1] = np.inf
pd.cut(x, bins)
```

**Comment by jreback** [comment]
@dberenbaum so I this for ``.cut()`` this is not needed at all as a user can just specify ``np.inf`` as a bound, for ``.qcut()`` its between 0 and 1, the quantiles. So can you give a case where this is actually useful?

**Comment by dberenbaum** [comment]
Please see #17282 for an example of how this is useful. That example is very similar to the cases where I would have found this useful. Does that example clarify the intended usage?

**Comment by jreback** [comment]
@dberenbaum did you see my comment. ``qcut`` is for ``quantile`` cutting, not arbitrary bounds, where ``cut`` suffices.

**Comment by dberenbaum** [comment]
@jreback I saw your comment. Did you look at the example in #17282?

Let's say I use `qcut` to bin a sample of data collected into quartiles:

```
>>>q = pd.qcut([1,2,3,4,5,6,7,8,9], 4)
>>>q
[(0.999, 3.0], (0.999, 3.0], (0.999, 3.0], (3.0, 5.0], (3.0, 5.0], (5.0, 7.0], (5.0, 7.0], (7.0, 9.0], (7.0, 9.0]]
Categories (4, interval[float64]): [(0.999, 3.0] < (3.0, 5.0] < (5.0, 7.0] < (7.0, 9.0]]
```

I subsequently collect a new data sample, which contains values > 9. I want to compare the new sample to my original data using the returned quartiles. There is no quartile for the new values > 9. With the `bounded` option, I could do the following:

```
>>>q = pd.qcut([1,2,3,4,5,6,7,8,9], 4, bounded=False)
>>>q
[(-np.inf, 3.0], (0.999, 3.0], (0.999, 3.0], (3.0, 5.0], (3.0, 5.0], (5.0, 7.0], (5.0, 7.0], (7.0, 9.0], (7.0, np.inf]]
Categories (4, interval[float64]): [(-np.inf, 3.0] < (3.0, 5.0] < (5.0, 7.0] < (7.0, np.inf]]
```

Now, I have useful quartiles for any subsequent code that uses the output of `qcut`, even if I'm applying those quartiles to new data that falls outside the range of the original sample.

The `bounded` behavior might not always be desirable, but I've found it to at least be a useful option.

**Comment by jreback** [comment]
@dberenbaum ok I looked at this again and it is ok, can you merge master

File: pandas/tests/reshape/test_qcut.py [hunk 4]
```diff
@@ -197,3 +197,30 @@ def test_date_like_qcut_bins(arg, expected_bins):
     ser = Series(arg)
     result, result_bins = qcut(ser, 2, retbins=True)
     tm.assert_index_equal(result_bins, expected_bins)
+
+
+def test_qcut_unbounded():
```
Comment (jreback) [inline pandas/tests/reshape/test_qcut.py]: can you parametrize over bounded

File: pandas/tests/reshape/test_qcut.py [hunk 5]
```diff
@@ -197,3 +197,30 @@ def test_date_like_qcut_bins(arg, expected_bins):
...
+    # GH 17282
+    labels = qcut(range(5), 4, bounded=False)
+    left = labels.categories.left.values
+    right = labels.categories.right.values
+    expected = np.array([-np.inf, 1.0, 2.0, 3.0, np.inf])
```
Comment (jreback) [inline pandas/tests/reshape/test_qcut.py]: rather than use numpy arrays, can you construct the expected Index and use ``tm.assert_index_equal``

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.rst [hunk 6]
```diff
@@ -421,6 +421,7 @@ Other Enhancements
 - :func:`pandas.DataFrame.to_sql` has gained the ``method`` argument to control SQL insertion clause. See the :ref:`insertion method <io.sql.method>` section in the documentation. (:issue:`8953`)
 - :meth:`DataFrame.corrwith` now supports Spearman's rank correlation, Kendall's tau as well as callable correlation methods. (:issue:`21925`)
 - :meth:`DataFrame.to_json`, :meth:`DataFrame.to_csv`, :meth:`DataFrame.to_pickle`, and :meth:`DataFrame.to_XXX` etc. now support tilde(~) in path argument. (:issue:`23473`)
+- :func: qcut now accepts ``bounded`` as a keyword argument, allowing for unbounded quantiles such that the lower/upper bounds are -inf/inf (:issue:`17282`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.rst]: Move to 0.25 at this point

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you merge master?

**Comment by dberenbaum** [comment]
Closing this PR per my comment above.

**Pull request closed without merging**